### PR TITLE
Implementation till signin so far

### PR DESCRIPTION
**Pull request by xenowits**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -46,12 +50,29 @@ func WithError(next func(w http.ResponseWriter, r *http.Request) (int, error)) f
...
+	nonce := hex.EncodeToString(nonceBytes)
+	return nonce, nil
+}
+
 func RequestNonceHandler() func(w http.ResponseWriter, r *http.Request) (int, error) {
 	fn := func(w http.ResponseWriter, r *http.Request) (int, error) {
-		return http.StatusNotImplemented, terror.ErrNotImplemented
+		nonce, err := GenerateNonce()
+		println(nonce, err)
+		w.Write([]byte(nonce))
```
Comment (nii236) [inline main.go]: I'd like a more userfriendly message to be signed as a Typed Message:

image

```
// Example payload fields
type Payload struct {
    NonceHex string `json:"nonceHex"`
    Message string `json:"message"`
    Address string `json:"address"`
}

func GenerateNonce(message) (*Payload, error) {
    ...
}
```

This way we can make the UX a little better and readable to not scare away users.

File: signer.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        Version: "1",
+        Salt:    salt,
+        ChainId: big.NewInt(1),
+    },
+    Message: signer.TypedDataMessage{
+        "timestamp": timestamp,
+        "address":   walletAddress,
+        "nonce":     nonce,
+    },
+}
```
Comment (nii236) [inline signer.go]: Add new lines to end fo files. Relevant VSCode settings: `"files.insertFinalNewline": true`

File: main.go [hunk 3]
```diff
@@ -18,15 +20,17 @@ import (
 	"github.com/go-chi/chi/v5"
 )
 
-//go:embed web
-var web embed.FS
+////go:embed web
+//var web embed.FS
```
Comment (nii236) [inline main.go]: Just remove this if you don't need it

File: main.go [hunk 4]
```diff
@@ -46,12 +50,29 @@ func WithError(next func(w http.ResponseWriter, r *http.Request) (int, error)) f
...
+	}
+	nonce := hex.EncodeToString(nonceBytes)
+	return nonce, nil
+}
+
 func RequestNonceHandler() func(w http.ResponseWriter, r *http.Request) (int, error) {
 	fn := func(w http.ResponseWriter, r *http.Request) (int, error) {
-		return http.StatusNotImplemented, terror.ErrNotImplemented
+		nonce, err := GenerateNonce()
+		println(nonce, err)
```
Comment (nii236) [inline main.go]: Remove redundant log

File: main.go [hunk 5]
```diff
@@ -46,12 +50,29 @@ func WithError(next func(w http.ResponseWriter, r *http.Request) (int, error)) f
...
+		return "", err
+	}
+	nonce := hex.EncodeToString(nonceBytes)
+	return nonce, nil
+}
+
 func RequestNonceHandler() func(w http.ResponseWriter, r *http.Request) (int, error) {
 	fn := func(w http.ResponseWriter, r *http.Request) (int, error) {
-		return http.StatusNotImplemented, terror.ErrNotImplemented
+		nonce, err := GenerateNonce()
```
Comment (nii236) [inline main.go]: Handle error immediately

File: static/index.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      } else {
+        document.getElementById("sign").value = signature;
+      }
+    }
+  );
+};
+
+const verifySignature = async () => {
+  // TO BE IMPLEMENTED
+};
```
Comment (nii236) [inline static/index.js]: Newline here as well

File: static/index.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  // set publicAdress
+  var publicAddress = await getMetamaskAddress();
+  console.log(publicAddress);
+  document.getElementById("address").value = publicAddress;
+};
+
+const getMetamaskAddress = async () => {
+  // Log all eth accounts
+  const accounts = await ethereum.request({ method: "eth_accounts" });
+  console.log("Accounts are", accounts);
```
Comment (nii236) [inline static/index.js]: Remove redundant console logs

**Review by nii236 — CHANGES_REQUESTED** [review]
Great work @xenowits, some small changes requested, the main one the useof user friendly, human readable challenges.

**Comment by nii236** [comment]
Actually instead of JSON it would be great to have just a string and we can do cool stuff like this from zapper.fi:

image

File: main.go (hunk 3 above)
Comment (xenowits) [reply main.go]: ok

**Comment by xenowits** [comment]
Hi @nii236 . Thanks for the review!
Metamask SS4

I did something like that. Checkout the screenshot

**Comment by xenowits** [comment]
Is this something that we want?

**Pull request merged**